### PR TITLE
Optimize `box + isinst + unbox.any` as `nop`

### PR DESCRIPTION
**Pull request by EgorBo**
This pattern may appear in pattern matching + generics, e.g.:
```csharp
int Test1<T>(T o)
{
    if (o is int x)
        return x;
    return 0;
}


// or


int Test2<T>(T o) => o is int n ? n : 42;


// or


int Test3<T>(T o)
{
    return o switch
    {
        int n => n,
        string str => str.Length,
        _ => 0
    };
}
```
see IL in [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAIBlACwEMoAHAGReADoAlAK4A7DNgC2MANxUqhAMx0SdAMJ0A3lTra6WnXu0BtAFLYMAcRjCYUbGAAUGAJ7sYEAGb3sogJQ+AugZ03hh0ACowAM4YpAA8YQB89mF0ED5BmpQ62cHudPYQwZHBonQI6Vk5VYQA7GUyldV1NA3ZAL6yjXQA9N2pUJ3ZJmaW1rYOzq4eXr4BQSHhURjE8UkpaXQAvAmpRSWhwnQA/HSHIHQoxA1Bvf2DOsMWVjZ2ji5uniF+gV0LEdHyVbJVIVbKZKraWq7SIAdzMYCYQTBSIhC0O21OaBR1VI9GiUC2O3xvE4VgA5hgmFiuhC6AB9Ql0GjYuhtVo6DqUNpAA===)

#### Current codegen (for `Test<int>`):
```asm
    L0000: push rsi
    L0001: sub rsp, 0x20
    L0005: mov esi, edx
    L0007: mov rcx, 0x7ffec352b1e8
    L0011: call 0x7fff23084690
    L0016: mov [rax+0x8], esi
    L0019: mov eax, [rax+0x8]
    L001c: add rsp, 0x20
    L0020: pop rsi
    L0021: ret
```
#### New codegen (for `Test<int>`):
```asm
    L0000: mov eax, edx
    L0002: ret
```

PS: it doesn't fix cases where interfaces/objects are used + inlining instead of generics -- these cases require more work

### CODE REVIEW

File: src/coreclr/src/jit/importer.cpp [hunk 1]
```diff
@@ -5941,6 +5941,28 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const B
                                 }
                             }
                         }
+                        break;
+
+                    // box + isinst + unbox.any
+                    case CEE_UNBOX_ANY:
+                        if ((nextCodeAddr + 5) <= codeEndp)
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/importer.cpp]: Probably clearer as `if (nextCodeAddr + 1 + sizeof(mdToken) <= codeEndp)` ?

File: src/coreclr/src/jit/importer.cpp [hunk 2]
```diff
@@ -5941,6 +5941,28 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const B
...
+                        {
+                            CORINFO_RESOLVED_TOKEN unboxResolvedToken = {};
+                            impResolveToken(codeAddr + 1, &unboxResolvedToken, CORINFO_TOKENKIND_Class);
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/importer.cpp]: For the unbox token, this should be `nextCodeAddr + 1`, right?

Shouldn't we be verifying that all 3 tokens in this sequence (box, isinst, unbox.any) must compare equal?

**Review by AndyAyersMS — COMMENTED** [review]
Would also be good to get some idea how often we see this in actual code -- any examples you can point at? Any hits in jit-diffs?

If there are no jit-diffs hits, please consider adding some test cases.

File: src/coreclr/src/jit/importer.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/src/jit/importer.cpp]: Good point!

**Comment by EgorBo** [comment]
@AndyAyersMS I've just added tests. The jit-diff is empty but maybe because this pattern was avoided.
I was implementing existing optimizations from `impBoxPatternMatch` in Mono runtime and noticed this case and decided to optimize since the fix takes just a several lines of code.

**Comment by tannergooding** [comment]
I imagine this pattern may become more common now that pattern matching exists (it is relatively new after all).

**Comment by EgorBo** [comment]
> I imagine this pattern may become more common now that pattern matching exists (it is relatively new after all).

Indeed!
E.g.:
```csharp
    static int Test<T>(T o)
    {
        return o switch
        {
            int n => n,
            string str => str.Length,
            _ => 0
        };
    }
```

File: src/coreclr/tests/src/JIT/Generics/Conversions/Boxing/box_isinst_unbox-noinline.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <DebugType>None</DebugType>
+    <Optimize>True</Optimize>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="box_isinst_unbox.cs" />
+  </ItemGroup>
+  <PropertyGroup>
+    <CLRTestBatchPreCommands><![CDATA[
```
Comment (AndyAyersMS) [inline src/coreclr/tests/src/JIT/Generics/Conversions/Boxing/box_isinst_unbox-noinline.csproj]: Would prefer to see this in source, via `[MethodImpl(MethodImplOptions.NoInlining)]`; does that not work for some reason?

File: src/coreclr/src/jit/importer.cpp [hunk 4]
```diff
@@ -5941,6 +5941,34 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const B
...
+                        break;
+
+                    // box + isinst + unbox.any
+                    case CEE_UNBOX_ANY:
+                        if ((nextCodeAddr + 1 + sizeof(mdToken)) <= codeEndp)
+                        {
+                            CORINFO_RESOLVED_TOKEN isinstResolvedToken = {};
+                            impResolveToken(codeAddr + 1, &isinstResolvedToken, CORINFO_TOKENKIND_Class);
+
+                            CORINFO_RESOLVED_TOKEN unboxResolvedToken = {};
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/importer.cpp]: The callbacks from the jit to the runtime have some cost.  Can you first check if isinst and box agree and only if that is so, resolve the unbox token and check it against the box.

File: src/coreclr/tests/src/JIT/Generics/Conversions/Boxing/box_isinst_unbox-noinline.csproj (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/tests/src/JIT/Generics/Conversions/Boxing/box_isinst_unbox-noinline.csproj]: @AndyAyersMS decided to make two `csproj` with and without inlining but can make only one + NoInlining where needed

**Comment by EgorBo** [comment]
Wonder if I should also (if can) handle
```
        IL_0001: box !!T
        IL_0006: isinst IFoo
        IL_000b: stloc.0
        IL_000c: ldloc.0
        IL_000d: brtrue.s IL_0012
```
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAIBlACwEMoAHAGReADoAlAK4A7DNgC2MANxUqhAMx0SdAMJ0A3lTra6WnQG0AUtgwBxGMJhRsYABQYAnuxgQAZrYBiECAEofAXT1tBTpsUToAFRgAZwxiAB4IgD5bCLpfOgBeJPTQ6LoASS8IOmE6AH5S3nMMAEFbHzoQOhRiGUoAX1lKOUUwjCtXFjAYQuKqTUodJT7wmvqfdq6e6kVYqEEwDDpiprHvCaCZ0PCWdumQ/rp5hqycs6oOoA==

**Review by AndyAyersMS — COMMENTED** [review]
Thanks for the test cases. These examples all get optimized and no longer allocate, right?

Should we try and verify there are no GC allocations done by these methods? See for instance `JIT\opt\ObjectStackAllocation\ObjectStackAllocationTests.cs`, though you shouldn't need to check if SPC is optimized or not.

And since this is an unconditional optimization (that is, we're optimizing these even when generating debuggable code), is it possible this might impair debuggability? I suppose C#'s debug IL ensures stack empty at each possible debug point, so probably not...

**Comment by AndyAyersMS** [comment]
The cases where the box gets stored (or dup'd) are more complex; you'll need to be able to see all the places that local is used, and prove the boxed value can reach those uses. Tricky pull off early, without dataflow...

**Comment by EgorBo** [comment]
> Thanks for the test cases. These examples all get optimized and no longer allocate, right?

Nope, some of them still allocate e.g. where Interfaces and Nullable are involved (`info.compCompHnd->compareTypesForEquality` gives up if one of the types is Nullable)

> And since this is an unconditional optimization (that is, we're optimizing these even when generating debuggable code), is it possible this might impair debuggability? I suppose C#'s debug IL ensures stack empty at each possible debug point, so probably not...

Should I hide the whole `importPatternMatch` under OptimizationsEnabled condition?

File: src/coreclr/src/jit/importer.cpp [hunk 5]
```diff
@@ -5947,21 +5947,21 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const B
...
                             impResolveToken(codeAddr + 1, &isinstResolvedToken, CORINFO_TOKENKIND_Class);
 
-                            CORINFO_RESOLVED_TOKEN unboxResolvedToken = {};
-                            impResolveToken(nextCodeAddr + 1, &unboxResolvedToken, CORINFO_TOKENKIND_Class);
-
-                            // See if the resolved tokens describe types that are equal.
                             const TypeCompareState compareBoxIsInst =
                                 info.compCompHnd->compareTypesForEquality(isinstResolvedToken.hClass,
                                                                           pResolvedToken->hClass);
+                            CORINFO_RESOLVED_TOKEN unboxResolvedToken = {};
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/importer.cpp]: This doesn't defer the cost of the second set of calls -- resolution and type compare for the unbox token should happen only if the isinst comparsion is successful.

**Comment by AndyAyersMS** [comment]
> Should I hide the whole impBoxPatternMatch under OptimizationsEnabled condition?

No. We want box opts to happen at Tier0, which (for now) implies no optimizations enabled.

**Comment by EgorBo** [comment]
@AndyAyersMS is there anything left to improve in this PR?

**Review by AndyAyersMS — APPROVED** [review]
Looks good. Thanks!

**Comment by Sergio0694** [comment]
Commenting here following a conversation with @tannergooding on the [UWP Community Discord Server](https://discord.gg/mvp5TTP), I was wondering whether this PR would also address some codegen other inefficiencies I've noticed or whether I should open a separate issue for them.
Asking here first to double check and avoid creating duplicate issues 😄

<details>
  <summary><b>object is T, when T is a struct</b> (click to expand)</summary>
<br/>

image

</details>

<details>
  <summary><b>object is T value, when T is a struct</b> (click to expand)</summary>
<br/>

image

</details>

<details>
  <summary><b>object is T, when T is a sealed class</b> (click to expand)</summary>
<br/>

image

</details>

<details>
  <summary><b>object is T value, when T is a sealed class</b> (click to expand)</summary>
<br/>

image

</details>

You can see that in all these 4 cases, it's possible to produce better codegen by doing checks manually. I might be missing something, but I think the JIT should be able to understand the context here and perform these optimizations automatically?

Posting just images here for brevity, if you confirm I should open a separate issue to track these I'll write the various C#/asm snippets properly with markdown code blocks of course 😊

_Bonus point: case **2)** could also use the `no.(nullcheck|typecheck)` opcode once it's supported, as that unboxing is perfectly safe there, we're already checking everything on our own._

**Comment by drieseng** [comment]
@Sergio0694 I may be better to create a new issue rather than add a comment to a merged PR.

**Comment by EgorBo** [comment]
@Sergio0694 not sure, had a quick glance and looks like the main issue here is how JIT rotates comparisons but aside that "slow" and "fast" look the same to me

**Comment by Sergio0694** [comment]
@EgorBo Thanks for taking a look, appreciate it! 😊

There are mainly two differences I'm seeing in all of them:
- One less conditional branch in the "fast" version
- Slightly smaller codegen (this might in part go away if the method is inlined though)

Consider the first example:

image

The "fast" version has a single branch just for the `null` check, then it simply does a compare with the method table pointer and sets the flag, whereas the other handles each condition separately 🤔

The worst offender would be the second example though, where the `(T)` checks seem completely redundant, but as mentioned above I guess we'll just have to wait for the runtime to support the `no.` prefix for that (and that seems like a separate issue as well).

**Comment by EgorBo** [comment]
@Sergio0694 feel free to file an issue. Ideally, with code, not the screenshots ;-)

**Comment by Sergio0694** [comment]
@EgorBo Absolutely, will do!

And yeah don't worry, the issue will have actual code, the screens were just to quickly put together this exploratory question to make sure I wasn't asking a dupe of something else 😄

**Pull request merged**